metric_learn/base_metric.py

```diff
@@ -396,6 +395,150 @@ def _set_default_threshold(self, pairs, y):
+
+    y_valid : array-like, shape=(n_pairs_valid,)
+      The labels of the pairs of the validation set to use to set the
+      threshold.
+
+    References
+    ----------
+    .. [1] Receiver-operating characteristic (ROC) plots: a fundamental
+           evaluation tool in clinical medicine, MH Zweig, G Campbell -
+           Clinical chemistry, 1993
+
+    .. [2] most of the code of this function is from scikit-learn's PR #10117
+
+    See Also
+    --------
+    sklearn.calibration : scikit-learn's module for calibrating classifiers
+    """
+
+    if strategy not in ('accuracy', 'f_beta', 'max_tpr',
+                        'max_tnr'):
+      raise ValueError('Strategy can either be "accuracy", "f_beta" or '
+                       '"max_tpr" or "max_tnr". Got "{}" instead.'
+                       .format(strategy))
+
+    if strategy == 'max_tpr' or strategy == 'max_tnr':
+      if (threshold is None or not isinstance(threshold, (int, float)) or
+          not threshold >= 0 or not threshold <= 1):
+        raise ValueError('Parameter threshold must be a number in'
+                         '[0, 1]. '
+                         'Got {} instead.'.format(threshold))
+
+    if strategy == 'f_beta':
+      if beta is None or not isinstance(beta, (int, float)):
+        raise ValueError('Parameter beta must be a real number. '
+                         'Got {} instead.'.format(type(beta)))
+
+    pairs_valid, y_valid = self._prepare_inputs(pairs_valid, y_valid,
+                                                type_of_inputs='tuples')
+
+    n_samples = pairs_valid.shape[0]
+    if strategy == 'accuracy':
+      scores = self.decision_function(pairs_valid)
+      scores_sorted_idces = np.argsort(scores)[::-1]
+      scores_sorted = scores[scores_sorted_idces]
+      # true labels ordered by decision_function value: (higher first)
+      y_ordered = y_valid[scores_sorted_idces]
+      # finds the threshold that maximizes the accuracy:
+      cum_tp = stable_cumsum(y_ordered == 1)  # cumulative number of true
+      # positives
+      cum_tn_inverted = stable_cumsum(y_ordered[::-1] == -1)
+      cum_tn = np.concatenate([[0], cum_tn_inverted[:-1]])[::-1]
+      cum_accuracy = (cum_tp + cum_tn) / n_samples
+      max_i = np.argmax(cum_accuracy)
+      # note: we want a positive threshold (distance), so we take - threshold
+      self.threshold_ = - scores_sorted[max_i]
+      return self
+
+    if strategy == 'f_beta':
+      precision, recall, thresholds = precision_recall_curve(
+          y_valid, self.decision_function(pairs_valid), pos_label=1)
+      with np.errstate(divide='ignore', invalid='ignore'):
+        f_beta = ((1 + beta**2) * (precision * recall) /
+                  (beta**2 * precision + recall))
+      f_beta[np.isnan(f_beta)] = 0.
+      imax = np.argmax(f_beta)
+      self.threshold_ = - thresholds[imax]
+      return self
+
+    fpr, tpr, thresholds = roc_curve(y_valid,
```

Comment:
Note that there is a particular corner case that is not explicitely coded here but comes from scikit-learn: we should decide if that's what we want or not: 
In `roc_curve`, an additional point is added in (0, 0) (0 false positives and 0 true positives (so 0 positives)) if needed (i.e. if there is not already such a point), with an artificial threshold which is thresholds[0] + 1 (see  https://github.com/scikit-learn/scikit-learn/blob/62d205980446a1abc1065f4332fd74eee57fcf73/sklearn/metrics/ranking.py#L637-L642). (this points represent a case where we want to reject all samples, the +1 makes sense since it will reject all samples from the training set)
Is that what we want ? It kind of makes sense since it works if we want to reject everything but the +1 seems a bit arbitrary

Reply:
Note also that some points can be deleted in this curve (see these lines of code, and PR #5237). But it will not change the result as it just removes suboptimal points (and keeps optimal points i.e. points with the highest tpr for the same tnr and points with the highest tnr for the same tpr). We could have avoided this deletion by setting `drop_intermediate=True` but we don't need to do that since the deletion is harmless.

Reply:
I would say that it is fine this way? It does not seem very important anyway so going with the sklearn approach is fine